### ISSUE TITLE
Not compatible with Zombie 4.x

### ISSUE THREAD

**Issue by jhedstrom** [minkphp/MinkZombieDriver#132]
When running against the latest Zombie, I am getting the following:

```
    Given I am on the homepage                         # Drupal\DrupalExtension\Context\MinkContext::iAmOnHomepage()
      Server process has been terminated: (8) [
      /usr/local/lib/node_modules/zombie/node_modules/jsdom/lib/jsdom/level2/html.js:238
      var nonInheritedTags = new Set([
                                 ^
      ReferenceError: Set is not defined
          at Object.<anonymous> (/usr/local/lib/node_modules/zombie/node_modules/jsdom/lib/jsdom/level2/html.js:238:28)
          at Module._compile (module.js:456:26)
          at Object.Module._extensions..js (module.js:474:10)
          at Module.load (module.js:356:32)
          at Function.Module._load (module.js:312:12)
          at Module.require (module.js:364:17)
          at require (module.js:380:17)
          at Object.<anonymous> (/usr/local/lib/node_modules/zombie/node_modules/jsdom/lib/jsdom/living/index.js:8:1)
          at Module._compile (module.js:456:26)
          at Object.Module._extensions..js (module.js:474:10)
          at Module.load (module.js:356:32)
          at Function.Module._load (module.js:312:12)
          at Module.require (module.js:364:17)
          at require (module.js:380:17)
          at Object.<anonymous> (/usr/local/lib/node_modules/zombie/lib/dom/index.js:5:11)
          at Module._compile (module.js:456:26)
      ] (RuntimeException)
    When I press on "Commits"
```

**Comment by MKorostoff** [comment]
I'm seeing this issue too

**Comment by aik099** [comment]
Please provide:

# example page html, that fails
# debug output of Behat, while it runs scenario that triggers the exception 

Most likely we'll end up with a bug in Zombie or JSDOM itself.

**Comment by jhedstrom** [comment]
Sure.

`composer.json`:

```json
{
    "require": {
        "behat/mink-zombie-driver": "^1.2",
        "behat/behat": "^3.0",
        "behat/mink-extension": "^2.0"
    }
}
```

`behat.yml`:
```yaml
default:
  suites:
    default:
      contexts:
        - Behat\MinkExtension\Context\MinkContext
  extensions:
    Behat\MinkExtension:
      zombie: ~
      base_url: http://www.wikipedia.org
      javascript_session: zombie
```

`features/test.feature`:
```cucumber
Feature: Illustrate Zombie JS behavior.

  @javascript
  Scenario: Visit Wikipedia
    Given I am on the homepage
```

**Comment by stof** [comment]
your issue with Set not being defined is an issue in your setup itself: Zombie 4 is not compatible with node.js 0.12 anymore. It requires running on io.js. So you would need to configure the driver to find io.js (or symlink it to make it available as ``node`` in your path):

```yaml
default:
  suites:
    default:
      contexts:
        - Behat\MinkExtension\Context\MinkContext
  extensions:
    Behat\MinkExtension:
      zombie:
        node_bin: iojs
      base_url: http://www.wikipedia.org
      javascript_session: zombie
```

Note however that we have not tested the driver on Zombie 4 yet (I thought about doing it but Travis does not provide io.js on PHP VMs, making it much more painful)

**Comment by jhedstrom** [comment]
Great, thanks @stof!

**Comment by stof** [comment]
@jhedstrom if you have an io.js setup, I would be interested in having a testsuite output on Zombie 4 btw. Here is what would be needed:

- clone the MinkZombieDriver repo
- install the dependencies with ``composer install``
- install Zombie with ``npm install zombie``
- start a server for the testsuite web-fixtures: ``php -S localhost:8080 -t vendor/behat/mink/driver-testsuite/web-fixtures`` (assuming you have PHP 5.4+, use another server supporting PHP otherwise)
- export the needed environment variable to set the fixtures host (or copy the ``phpunit.xml.dist`` to ``phpunit.xml`` and uncomment the settings configuring them):
    - ``export WEB_FIXTURES_HOST='http://localhost:8080'``
- change the node bin in the testsuite in https://github.com/minkphp/MinkZombieDriver/blob/db0ae2f95b8b0e9ea7c5a3ab44f28da112bb08ea/tests/ZombieConfig.php#L20 to use iojs (we don't yet have a way to make it configurable)
- run ``phpunit -v`` at the root of the package
- paste us the output (in a gist as it is more readable than in a github comment).

This would help us seeing how much we actually support Zombie 4 (I don't have an io.js setup locally to test it)

**Comment by jhedstrom** [comment]
@stof I don't currently have io.js running locally either (I've simply downgraded zombie back to ~2.5 for now). I'll report back here when I do though.

Re-opening the issue so I don't forget.

**Comment by aik099** [comment]
I've downloaded io.js binary and tests do fail too with following output (it takes too long to fail, so I just executed them until 1st failure):

```
phpunit --stop-on-failure
PHPUnit 4.6.10 by Sebastian Bergmann and contributors.

Configuration read from .../phpunit.xml

E

Time: 6 seconds, Memory: 9.25Mb

There was 1 error:

1) Behat\Mink\Tests\Driver\Custom\ExtraTest::testSetUserAgent
Behat\Mink\Exception\DriverException: Error "Cannot read property 'nodeType' of null" while executing code: var node,
    refs = [],
    result = browser.xpath("\/\/html");

while (node = result.iterateNext()) {
    if (node.nodeType !== 10) {
        pointers.push(node);
        refs.push(pointers.length - 1);
    }
}
stream.end(JSON.stringify(refs));

.../src/NodeJS/Server/ZombieServer.php:44
.../src/NodeJS/Server.php:366
.../src/ZombieDriver.php:341
.../vendor/behat/mink/src/Driver/CoreDriver.php:101
.../src/ZombieDriver.php:922
.../src/ZombieDriver.php:377
.../vendor/behat/mink/src/Element/Element.php:172
.../tests/Custom/ExtraTest.php:19

FAILURES!
Tests: 1, Assertions: 0, Errors: 1.
```

**Comment by stof** [comment]
Wait, ``null`` should stop the while loop here. How is it possible to have the inner code executed with ``node = null`` ?

**Comment by aik099** [comment]
Ah, that was me. I've moved the driver repo clone on disk, but haven't changed it's path in `phpunit.xml` resulting in Zombie going to page, that returned 404 all the time hence the error. But indeed it's strange that nothing was found. Maybe that 404 page hadn't `<html>` tag on it at all.

The correct output from running on io.js is this:

```
PHPUnit 4.6.10 by Sebastian Bergmann and contributors.

Configuration read from .../phpunit.xml

...............................................................  63 / 166 ( 37%)
...FS...S......SSSSS...........F.F.SSSS...............S........ 126 / 166 ( 75%)
.......F.........S...............S..SSS

Time: 1.04 minutes, Memory: 12.50Mb

There were 4 failures:

1) Behat\Mink\Tests\Driver\Basic\HeaderTest::testResponseHeaders
Failed asserting that an array has the key 'x-mink-test'.

.../vendor/behat/mink/driver-testsuite/tests/Basic/HeaderTest.php:75

2) Behat\Mink\Tests\Driver\Form\GeneralTest::testAdvancedForm
Failed asserting that '<!DOCTYPE html>
<html>
<head>
    <title>Advanced form save</title>
    <meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
</head>
<body>
array (
  'agreement' = 'on',
  'email' = 'Content-Length: 18

ever.zet@gmail.com',
  'first_name' = 'Content-Length: 10

Foo "item"',
  'last_name' = 'Content-Length: 3

Bar',
  'notes' = 'Content-Length: 9

new notes',
  'select_number' = 'Content-Length: 2

30',
  'sex' = 'Content-Length: 1

m',
  'submit' = 'Content-Length: 8

Register',
)
some_file.txt
1 uploaded file
</body>
</html>' contains "array (
  'agreement' = 'on',
  'email' = 'ever.zet@gmail.com',
  'first_name' = 'Foo "item"',
  'last_name' = 'Bar',
  'notes' = 'new notes',
  'select_number' = '30',
  'sex' = 'm',
  'submit' = 'Register',
)
some_file.txt
1 uploaded file".

.../vendor/behat/mink/driver-testsuite/tests/Form/GeneralTest.php:220

3) Behat\Mink\Tests\Driver\Form\GeneralTest::testAdvancedFormSecondSubmit
Failed asserting that '<!DOCTYPE html>
<html>
<head>
    <title>Advanced form save</title>
    <meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
</head>
<body>
array (
  'agreement' = 'off',
  'email' = 'Content-Length: 14

your@email.com',
  'first_name' = 'Content-Length: 9

Firstname',
  'last_name' = 'Content-Length: 8

Lastname',
  'mail_list' = 'Content-Length: 2

on',
  'notes' = 'Content-Length: 14

original notes',
  'select_number' = 'Content-Length: 2

20',
  'sex' = 'Content-Length: 1

w',
  'submit' = 'Content-Length: 5

Login',
)
no file</body>
</html>' contains "'submit' = 'Login',".

.../vendor/behat/mink/driver-testsuite/tests/Form/GeneralTest.php:289

4) Behat\Mink\Tests\Driver\Js\EventsTest::testRightClick
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'right clicked'
+'not clicked'

.../vendor/behat/mink/driver-testsuite/tests/Js/EventsTest.php:45

FAILURES!
Tests: 163, Assertions: 393, Failures: 4, Skipped: 17.
```

**Comment by stof** [comment]
Looks like their multipart submission is broken. this should probably be reported as a bug to Zombie

**Comment by aik099** [comment]
I guess so. Please create one.

I wonder how this issue (multipart form submit) wasn't noticed by any users of Zombie 4.x release.

**Comment by stof** [comment]
reported. I think I even found the issue (you could even try whether it really fixes it by removing a CRLF on the line linked in the zombie ticket and running the ``Form\GeneralTest::testAdvancedForm`` again

**Comment by stof** [comment]
even better: try with Zombie 4.0.13 as the fix has been released

**Comment by aik099** [comment]
It works now, but another test started to fail, that passed before:

```
PHPUnit 4.6.10 by Sebastian Bergmann and contributors.

Configuration read from .../phpunit.xml

...............................................................  63 / 166 ( 37%)
...FS...S......SSSSS...............SSSS...............S........ 126 / 166 ( 75%)
.................S...............S..SSS

Time: 19.64 seconds, Memory: 12.50Mb

There was 1 failure:

1) Behat\Mink\Tests\Driver\Basic\HeaderTest::testResponseHeaders
Failed asserting that an array has the key 'x-mink-test'.

.../vendor/behat/mink/driver-testsuite/tests/Basic/HeaderTest.php:75

FAILURES!
Tests: 163, Assertions: 394, Failures: 1, Skipped: 17.
```

P.S.
I think, that we should note in README, that:

1. Zombie 4.x+ requires IO.JS to run instead of Node.JS
2. versions of Zombie > 4.x and < 4.0.12 doesn't support multi-part form submission

**Comment by aik099** [comment]
That `testResponseHeaders` test fails because the `browser.window._response.headers` now evaluates to `_headers` array which in turn contains actual headers. Before it was an array of headers.

Not sure if the change in Zombie to do this was intentional or a bug.

**Comment by stof** [comment]
@aik099 properties prefixed with underscore are generally considered internal ones in JS library, so it may have be seen as acceptable

**Comment by aik099** [comment]
You mean `not acceptable`? Should we use properties starting with `_` or not? Anyway in #134 I've used `.toObject()` method of `headers` property to get actual headers instead of getting it from `_headers` private property.

**Comment by stof** [comment]
it may have been seen as acceptable to change the way to access headers, as there was no public API to access them before (and btw, BC breaks are allowed in major versions anyway)

**Comment by aik099** [comment]
Closing, because Zombie 4.x related issues were fixed in #136, #137 and #138 PRs.
